File: Data-Structures/02_Linked_List/02_Doubly_Linked_List.cpp
#include <iostream>

class ListNode {
public:
    int val;
    ListNode* prev = nullptr;
    ListNode* next = nullptr;

    // Constructors
    ListNode() {}
    ListNode(int val) {
        this->val = val;
    }
};

class LinkedList {
private:
    ListNode* head = nullptr;
    ListNode* tail = nullptr;
public:
    // Constructor
    LinkedList() {}

    // Check if doubly linked list is empty
    bool isEmpty() {
        if (head == nullptr) {
            return true;
        }

        return false;
    }

    // Insert node at the head of the doubly linked list
    void insertFront(int val) {
        ListNode* new_node = new ListNode(val);

        if (isEmpty()) {
            head = new_node;
            tail = head;
        } else {
            new_node->next = head;
            head->prev = new_node;
            head = new_node;
        }
    }

    // Insert node at the tail of the doubly linked list
    void insertEnd(int val) {
        ListNode* new_node = new ListNode(val);

        if (isEmpty()) {
            tail = new_node;
            head = tail;
        } else {
            new_node->prev = tail;
            tail->next = new_node;
            tail = new_node;
        }
    }

    // Remove node at the head of the doubly linked list
    void removeFront() {
        if (!isEmpty()) {
            if (head == tail) {
                head = nullptr;
                tail = nullptr;
            } else {
                head->next->prev = nullptr;
                head = head->next;
            }
        }
    }

    // Remove node at the tail of the doubly linked list
    void removeEnd() {
        if (!isEmpty()) {
            if (head == tail) {
                tail = nullptr;
                head = nullptr;
            } else {
                tail->prev->next = nullptr;
                tail = tail->prev;
            }
        }
    }

    // Print the doubly linked list
    void print() {
        ListNode* curr = head;
        while (curr != nullptr) {
            printf("<-[%d]->", curr->val);
            curr = curr->next;
        }
        printf("\n");
    }
};

int main() {
    LinkedList dll1;

    printf("insertFront(2), insertEnd(3), insertFront(1), print(): \n");
    dll1.insertFront(2);
    dll1.insertEnd(3);
    dll1.insertFront(1);
    dll1.print();

    printf("removeFront(), removeEnd(), print(): \n");
    dll1.removeFront();
    dll1.removeEnd();
    dll1.print();

    return 0;
}

File: Data-Structures/02_Linked_List/02_Doubly_Linked_List.py
class ListNode:
    def __init__(self, val = None, prev = None, next = None):
        self.val = val
        self.prev = prev
        self.next = next

class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
    
    # Check if doubly linked list is empty
    def isEmpty(self):
        if self.head is None:
            return True
        
        return False
    
    # Insert node at the head of the doubly linked list
    def insertFront(self, val):
        new_node = ListNode(val)

        if self.isEmpty():
            self.head = new_node
            self.tail = self.head
        else:
            new_node.next = self.head
            self.head.prev = new_node
            self.head = new_node
    
    # Insert node at the tail of the doubly linked list
    def insertEnd(self, val):
        new_node = ListNode(val)

        if self.isEmpty():
            self.tail = new_node
            self.head = self.tail
        else:
            new_node.prev = self.tail
            self.tail.next = new_node
            self.tail = new_node
    
    # Remove node at the head of the doubly linked list
    def removeFront(self):
        if self.isEmpty():
            return
        
        if self.head == self.tail:
            self.head = None
            self.tail = None
        else:
            self.head.next.prev = None
            self.head = self.head.next
    
    # Remove node at the tail of the doubly linked list
    def removeEnd(self):
        if self.isEmpty():
            return
        
        if self.head == self.tail:
            self.tail = None
            self.head = None
        else:
            self.tail.prev.next = None
            self.tail = self.tail.prev
    
    def getValues(self):
        values = []
        curr = self.head

        while curr:
            values.append(curr.val)
            curr = curr.next
        
        return values

def main():
    DLL1 = LinkedList()

    print("insertFront(2), insertEnd(3), insertFront(1), print():")
    DLL1.insertFront(2)
    DLL1.insertEnd(3)
    DLL1.insertFront(1)
    print(DLL1.getValues())

    print("removeFront(), removeEnd(), print():")
    DLL1.removeFront()
    DLL1.removeEnd()
    print(DLL1.getValues())

if __name__ == "__main__":
    main()